File: DSA-Sheets/Striver-SDE-Sheet/02-Arr-II/Find missing and repeating number/xor.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> repeatedNumber(const vector<int> &A)
{
    long long n = A.size();
    int xr = 0;
    for (int i = 0; i < n; i++)
    {
        xr = xr ^ A[i];  // for xoring with element in arr
        xr = xr ^ i + 1; // for xor with natural number
    }

    // find the last uncommon bit for differentitating elements
    int bitNo = 0;
    while (true)
    {
        if (xr & (1 << bitNo) != 0)
            break;
        bitNo++;
    }

    int zeros = 0;
    int ones = 0;
    for (int i = 0; i < n; i++)
    {
        // putting element in clubbing
        if (A[i] & (1 << bitNo))
            ones = ones ^ A[i];
        else
            zeros = zeros ^ A[i];

        // putting Number in clubbing
        if ((i + 1) & (1 << bitNo))
            ones = ones ^ (i + 1);
        else
            zeros = zeros ^ (i + 1);
    }

    // confirm which is repeating and which is missing
    int cnt = 0;
    for (auto itr : A)
        if (itr == zeros)
            cnt++;

    if (cnt == 2)
        return {zeros, ones};
    return {ones, zeros};
}

File: DSA-Sheets/Striver-SDE-Sheet/02-Arr-II/Find missing and repeating number/xor.py



from typing import List

def findMissingRepeatingNumbers(a: [int]) -> [int]:
    n = len(a) # size of the array

    xr = 0

    #Step 1: Find XOR of all elements:
    for i in range(n):
        xr = xr ^ a[i]
        xr = xr ^ (i + 1)

    #Step 2: Find the differentiating bit number:
    number = (xr & ~(xr - 1))

    #Step 3: Group the numbers:
    zero = 0
    one = 0
    for i in range(n):
        #part of 1 group:
        if ((a[i] & number) != 0):
            one = one ^ a[i]
        #part of 0 group:
        else:
            zero = zero ^ a[i]

    for i in range(1, n + 1):
        #part of 1 group:
        if ((i & number) != 0):
            one = one ^ i
        #part of 0 group:
        else:
            zero = zero ^ i

    # Last step: Identify the numbers:
    cnt = 0
    for i in range(n):
        if (a[i] == zero):
            cnt += 1

    if (cnt == 2):
        return [zero, one]
    return [one, zero]

if __name__ == '__main__':
    a = [3, 1, 2, 5, 4, 6, 7, 5]
    ans = findMissingRepeatingNumbers(a)
    print("The repeating and missing numbers are: {", ans[0], ", ", ans[1], "}\n") 

